File: src/Functional/Scala/Flyweight.scala
import scala.collection.mutable

final case class TextStyle(font: String, size: Int, color: String)

final class StyleFactory:
  private val pool = mutable.Map.empty[(String, Int, String), TextStyle]

  def get(font: String, size: Int, color: String): TextStyle =
    pool.getOrElseUpdate((font, size, color), TextStyle(font, size, color))

  def count: Int = pool.size

@main def flyweightExample(): Unit =
  val styles = StyleFactory()
  val red1 = styles.get("Inter", 12, "red")
  val red2 = styles.get("Inter", 12, "red")
  val blue = styles.get("Inter", 12, "blue")
  require(blue.color == "blue")
  val shared = red1 eq red2
  println(s"styles=${styles.count};shared=$shared;text=ABC")

File: src/Functional/Haskell/flyweight.hs
import qualified Data.Map.Strict as M

data TextStyle = TextStyle String Int String deriving (Eq, Show)
type Pool = M.Map (String, Int, String) TextStyle

intern :: Pool -> (String, Int, String) -> (TextStyle, Pool)
intern pool key@(font, size, color) =
  case M.lookup key pool of
    Just style -> (style, pool)
    Nothing -> let style = TextStyle font size color in (style, M.insert key style pool)

main :: IO ()
main = do
  let (red1, p1) = intern M.empty ("Inter", 12, "red")
      (red2, p2) = intern p1 ("Inter", 12, "red")
      (_, p3) = intern p2 ("Inter", 12, "blue")
      shared = red1 == red2
  putStrLn $ "styles=" ++ show (M.size p3) ++ ";shared=" ++ map toLowerBool (show shared) ++ ";text=ABC"
  where
    toLowerBool 'T' = 't'
    toLowerBool 'F' = 'f'
    toLowerBool c = c
